composer/core/data_spec.py

```diff
@@ -260,7 +261,14 @@ def _default_get_num_samples_in_batch(self, batch: Batch) -> int:
                     Please use a DataSpec and specify `get_num_samples_in_batch`."""))
 
     def _default_get_num_tokens_in_batch(self, batch: Batch) -> int:
-        del batch  # unused
+        # first try HuggingFace-style input dicts
+        if (isinstance(batch, dict) or isinstance(batch, UserDict)) and 'input_ids' in batch:
```

Comment:
why specifically UserDict? collections has a lot of types of Dicts, eg OrderedDict. Is there a more generic way to get these?

Reply:
perhaps i should use `Mapping`? I copied this line from a line I wrote in `huggingface.py`

```
In [11]: isinstance(ud, dict)
Out[11]: False

In [12]: isinstance(od, dict)
Out[12]: True

In [13]: isinstance(d, dict)
Out[13]: True

In [14]: isinstance(ud, Mapping)
Out[14]: True

In [15]: isinstance(od, Mapping)
Out[15]: True

In [16]: isinstance(d, Mapping)
Out[16]: True
```

Reply:
switch to mapping